### ISSUE TITLE
Functions with vars in their ref aren't disallowed at parse-/compile-time

### ISSUE THREAD

**Issue by johanfylling** [open-policy-agent/opa#5958]
The following module compiles without issue:

```rego
package test

p := a.b.c(2)

a.b[c](x) := x {c = "c"}
```

`a.b[c](x)` is however an "unreachable" function, as the var `c` is defined by calling and evaluating the function. Executing the policy will generate the error: `policy.rego:3: rego_type_error: undefined function data.test.a.b.c`

Even if we modify the module, adding a constant `c` that is know ahead of evaluation, evaluation will generate the same `rego_type_error`:

```rego
p := a.b.c(2)

a.b[c](x) := x 

c := "c"
```

## Expected behavior

Functions with vars defined in their ref should be disallowed at parse- or compile-time.

**Comment by ashutosh-narkar** [comment]
Just to understand this, OPA does generate the error at eval time and you're suggesting this should happen sooner? Are there cases where the policy needs to evaluated first to report such errors?

**Comment by johanfylling** [comment]
Correct. I have a hard time dreaming up a scenario where it would be beneficial to even have dynamic components in a function ref. Unless there's an argument not to, I think a simple rule to enforce at parse-time would be to require all function refs to be ground; i.e. no var components.

**Comment by anderseknert** [comment]
> I have a hard time dreaming up a scenario where it would be beneficial to even have dynamic components in a function ref

Dynamic policy composition, but with the inputs provided via args rather than via `with` and the caveats that comes with that. If this worked, it would be possible to have a main policy query any matching function dynamically, allowing functions to define dynamic requirements that could be added or removed by teams depending on their use case. More on that in [this issue](https://github.com/open-policy-agent/opa/issues/4025). Obviously, compile time checks of args and return values can't work for anything resolved at eval time, but I can think of some cases where that would be acceptable.
